### ISSUE TITLE
Numerical issue with Multinomial sampling

### ISSUE THREAD

**Issue by cossio** [JuliaStats/Distributions.jl#790]
There seems to be a numerical issue with Multinomial sampling. I think this was not occurring in previous versions of Distributions.jl.

I found this with a long probability vector. Here is the full source to reproduce this error:
https://gist.github.com/cossio/8c5bd3557aa1313919a3d0ccccef7ca2

This is on Distributions v0.16.4, Julia 1.0.1.

Here is the stack trace:

```
# ERROR: InexactError: Int64(Int64, NaN)
# Stacktrace:
#  [1] Type at ./float.jl:700 [inlined]
#  [2] convert at ./number.jl:7 [inlined]
#  [3] rand at /opt/julia-depot/packages/Distributions/WHjOk/src/univariate/discrete/binomial.jl:111 [inlined]
#  [4] multinom_rand!(::Int64, ::Array{Float64,1}, ::Array{Int64,1}) at /opt/julia-depot/packages/Distributions/WHjOk/src/samplers/multinomial.jl:14
#  [5] _rand! at /opt/julia-depot/packages/Distributions/WHjOk/src/multivariate/multinomial.jl:161 [inlined]
#  [6] rand(::Multinomial{Float64}) at /opt/julia-depot/packages/Distributions/WHjOk/src/multivariates.jl:49
#  [7] top-level scope at none:0
```

**Comment by simonbyrne** [comment]
Looks like an Rmath issue:
```
julia> using StatsFuns

julia> n = 4969027388018185
4969027388018185

julia> StatsFuns.RFunctions.binomrand(n, 0.1)
NaN
```

The solution is to switch over to our own samplers.

**Comment by cossio** [comment]
@simonbyrne Can you offer some guidance on how to fix this? I could then try to prepare a PR.

**Comment by simonbyrne** [comment]
There is some code here: https://github.com/JuliaStats/Distributions.jl/blob/master/src/samplers/binomial.jl
It needs testing and performance tuning.

**Comment by richardreeve** [comment]
This is now fixed by #830 using the sampler code @simonbyrne suggests (which is tested, but not performance tuned I'm afraid) and can be closed:
```julia
julia> include("Distributions_error.jl")
865-element Array{Int64,1}:
   222187754457
 10469767117680
  8688751380848
 11241013423469
  9679507254818
  7883122452775
  1852447899760
  1774469014382
      339509700
  6210597487262
  3058018576781
   336395441770
  8326451185448
  2009135287017
 11393951970349
 11207061922391
  2062490411856
              ⋮
 11372209007561
   695265175826
       79089195
  2400763729922
  4840171178085
 11420071634520
  2867397951915
  4300702751088
  9423816457952
 10988512108670
 11162656326794
  7923297021127
  8589241367401
    32437791018
 10923413779436
 10365920903760
```
